pkg/agent/prommetrics.go

```diff
@@ -16,52 +17,67 @@ type PromMetrics struct {
-			Help: "Number of panics from autoscaler-agent per-VM runner threads",
-		},
-	)
-	runnerStarts := prometheus.NewCounter(
-		prometheus.CounterOpts{
-			Name: "autoscaling_agent_runner_starts",
-			Help: "Number of new per-VM Runners started",
-		},
-	)
-	runnerRestarts := prometheus.NewCounter(
-		prometheus.CounterOpts{
-			Name: "autoscaling_agent_runner_restarts",
-			Help: "Number of existing per-VM Runners restarted due to failure",
-		},
-	)
-	totalErroredVMs := prometheus.NewGaugeFunc(
+	reg := prometheus.NewRegistry()
+
+	// register stock collectors directly:
+	//   (even though MustRegister is variadic, the function calls
+	//   are cheap and calling it more than once means that when
+	//   it panics, we know exactly which metric caused the error.)
+	reg.MustRegister(collectors.NewGoCollector())
+	reg.MustRegister(collectors.NewProcessCollector(collectors.ProcessCollectorOpts{}))
+
+	// the util.RegisterMetric() function registers the collector and returns
+	// it so we can set it directly on the output structure.
+	metrics := PromMetrics{
+		schedulerRequests: util.RegisterMetric(reg, prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_scheduler_plugin_requests_total",
+				Help: "Number of attempted HTTP requests to the scheduler plugin by autoscaler-agents",
+			},
+			[]string{"code"},
+		)),
+		informantRequestsOutbound: util.RegisterMetric(reg, prometheus.NewCounterVec(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_informant_outbound_requests_total",
+				Help: "Number of attempted HTTP requests to vm-informants by autoscaler-agents",
+			},
+			[]string{"code"},
+		)),
+		informantRequestsInbound: util.RegisterMetric(reg, prometheus.NewCounterVec(
+			prometheus.CounterOpts{Name: "autoscaling_agent_informant_inbound_requests_total",
+				Help: "Number of HTTP requests from vm-informants received by autoscaler-agents",
+			},
+			[]string{"endpoint", "code"},
+		)),
+		runnerFatalErrors: util.RegisterMetric(reg, prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_runner_fatal_errors_total",
+				Help: "Number of fatal errors from autoscaler-agent per-VM main runner thread",
+			},
+		)),
+		runnerThreadPanics: util.RegisterMetric(reg, prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_runner_thread_panics_total",
+				Help: "Number of panics from autoscaler-agent per-VM runner threads",
+			},
+		)),
+		runnerStarts: util.RegisterMetric(reg, prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_runner_starts",
+				Help: "Number of new per-VM Runners started",
+			},
+		)),
+		runnerRestarts: util.RegisterMetric(reg, prometheus.NewCounter(
+			prometheus.CounterOpts{
+				Name: "autoscaling_agent_runner_restarts",
+				Help: "Number of existing per-VM Runners restarted due to failure",
+			},
+		)),
+	}
+
+	// the remaining metrics are computed on timers run within
+	// prom: register them directly.
```

Comment:
I was under the impression that these are actually in response to metrics requests, not on some regular timer. Do you know if that's incorrect?

Reply:
huh, no idea where that idea came from on my part. you're right. I've modified my comment accordingly